src/log.cpp

```diff
@@ -19,195 +19,261 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-		/* LMT_INFO */    ANDROID_LOG_INFO,
-		/* LMT_VERBOSE */ ANDROID_LOG_VERBOSE
-	};
+static unsigned int level_to_android[] = {
+	ANDROID_LOG_INFO,     // LL_NONE
+	//ANDROID_LOG_FATAL,
+	ANDROID_LOG_ERROR,    // LL_ERROR
+	ANDROID_LOG_WARN,     // LL_WARNING
+	ANDROID_LOG_WARN,     // LL_ACTION
+	//ANDROID_LOG_INFO,
+	ANDROID_LOG_DEBUG,    // LL_INFO
+	ANDROID_LOG_VERBOSE,  // LL_VERBOSE
+
+};
 #endif
 
-std::vector<ILogOutput*> log_outputs[LMT_NUM_VALUES];
-std::map<threadid_t, std::string> log_threadnames;
-JMutex                            log_threadnamemutex;
 
-void log_add_output(ILogOutput *out, enum LogMessageLevel lev)
+LogLevel stringToLevel(const std::string &name)
 {
-	log_outputs[lev].push_back(out);
+	if      (name == "none")    return LL_NONE;
+	else if (name == "error")   return LL_ERROR;
+	else if (name == "warning") return LL_WARNING;
+	else if (name == "action")  return LL_ACTION;
+	else if (name == "info")    return LL_INFO;
+	else if (name == "verbose") return LL_VERBOSE;
+	else return LL_MAX;
 }
 
-void log_add_output_maxlev(ILogOutput *out, enum LogMessageLevel lev)
+
+void Logger::addOutput(ILogOutput *out, LogLevel lev)
 {
-	for(int i=0; i<=lev; i++)
-		log_outputs[i].push_back(out);
+	outputs[lev].push_back(out);
 }
 
-void log_add_output_all_levs(ILogOutput *out)
+void Logger::addOutputMaxLevel(ILogOutput *out, LogLevel lev)
 {
-	for(int i=0; i<LMT_NUM_VALUES; i++)
-		log_outputs[i].push_back(out);
+	for (unsigned i = 0; i <= lev; ++i) {
+		outputs[i].push_back(out);
+	}
 }
 
-void log_remove_output(ILogOutput *out)
+void Logger::removeOutput(ILogOutput *out)
 {
-	for(int i=0; i<LMT_NUM_VALUES; i++){
-		std::vector<ILogOutput*>::iterator it =
-				std::find(log_outputs[i].begin(), log_outputs[i].end(), out);
-		if(it != log_outputs[i].end())
-			log_outputs[i].erase(it);
+	for (unsigned i = 0; i < LL_MAX; ++i) {
+		OutputList::iterator it = std::find(
+				outputs[i].begin(), outputs[i].end(), out);
+		if (it != outputs[i].end())
+			outputs[i].erase(it);
 	}
 }
 
-void log_set_lev_silence(enum LogMessageLevel lev, bool silence)
+void Logger::setLevelSilenced(LogLevel lev, bool val)
 {
-	JMutexAutoLock lock(log_threadnamemutex);
-
-	for (std::vector<ILogOutput *>::iterator it = log_outputs[lev].begin();
-			it != log_outputs[lev].end(); ++it) {
-		ILogOutput *out = *it;
-		out->silence = silence;
-	}
+	silenced_levels[lev] = val;
 }
 
-void log_register_thread(const std::string &name)
+void Logger::registerThread(const std::string &name)
 {
 	threadid_t id = get_current_thread_id();
-	JMutexAutoLock lock(log_threadnamemutex);
-
-	log_threadnames[id] = name;
+	JMutexAutoLock lock(mutex);
+	thread_names[id] = name;
 }
 
-void log_deregister_thread()
+void Logger::deregisterThread()
 {
 	threadid_t id = get_current_thread_id();
-	JMutexAutoLock lock(log_threadnamemutex);
+	JMutexAutoLock lock(mutex);
+	thread_names.erase(id);
+}
 
-	log_threadnames.erase(id);
+const std::string Logger::getLevelName(LogLevel lev)
```

Comment:
Found the answer:
`getLevelName` returns a label for display when printing.  It should probably be named `getLevelLabel` since it doesn't actually convert to the level's name (eg, `LL_NONE` becomes `""`).  It's in the Logger class since it's tied to the actual logging and isn't useful outside of the logger.

`stringToLevel`, on the other hand, is a utility that's used to convert a level name to a level (eg, `"none"` -> `LL_NONE` and `""` -> error/`LL_MAX`).